File: myExpenses/src/main/java/org/totschnig/myexpenses/activity/MethodEdit.java
/*   This file is part of My Expenses.
 *   My Expenses is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   My Expenses is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with My Expenses.  If not, see <http://www.gnu.org/licenses/>.
*/

package org.totschnig.myexpenses.activity;

import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;

import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.model.AccountType;
import org.totschnig.myexpenses.model.Model;
import org.totschnig.myexpenses.model.PaymentMethod;
import org.totschnig.myexpenses.provider.DatabaseConstants;
import org.totschnig.myexpenses.ui.SpinnerHelper;

import java.util.NoSuchElementException;

import androidx.annotation.IdRes;
import androidx.appcompat.widget.AppCompatCheckBox;
import androidx.gridlayout.widget.GridLayout;

/**
 * Activity for editing an account
 * @author Michael Totschnig
 */
public class MethodEdit extends EditActivity implements CompoundButton.OnCheckedChangeListener {
  private EditText mLabelText;
  private GridLayout mAccountTypesGrid;
  private CheckBox mIsNumberedCheckBox;
  private SpinnerHelper mPaymentTypeSpinner;
  private PaymentMethod mMethod;

  @Override
  int getDiscardNewMessage() {
    return R.string.dialog_confirm_discard_new_method;
  }

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
        
    setContentView(R.layout.one_method);
    setupToolbar();

    mLabelText = findViewById(R.id.Label);
    mAccountTypesGrid = findViewById(R.id.AccountTypeGrid);

    mPaymentTypeSpinner = new SpinnerHelper(findViewById(R.id.TaType));

    mIsNumberedCheckBox = findViewById(R.id.IsNumbered);
    populateFields();
  }

  @Override
  protected void onResume() {
    super.onResume();
    setupListeners();
  }

  /**
   * populates the input field either from the database or with default value for currency (from Locale)
   */
  private void populateFields() {
    Bundle extras = getIntent().getExtras();
    long rowId = extras != null ? extras.getLong(DatabaseConstants.KEY_ROWID)
          : 0;
    int paymentType;
    if (rowId != 0) {
      mNewInstance = false;
      mMethod = PaymentMethod.getInstanceFromDb(rowId);

      setTitle(R.string.menu_edit_method);
      mLabelText.setText(mMethod.getLabel());
      paymentType = mMethod.getPaymentType();
      mIsNumberedCheckBox.setChecked(mMethod.isNumbered);
    } else {
      mMethod = new PaymentMethod();
      setTitle(R.string.menu_create_method);
      paymentType = PaymentMethod.NEUTRAL;
    }
    mPaymentTypeSpinner.setSelection(paymentType +1);
    //add one checkbox for each account type
    AppCompatCheckBox cb;
    for (AccountType accountType : AccountType.values()) {
      cb = new AppCompatCheckBox(this);
      cb.setText(accountType.toStringRes());
      cb.setTag(accountType);
      cb.setChecked(mMethod.isValidForAccountType(accountType));
      //setting Id makes state be retained on orientation change 
      cb.setId(getCheckBoxId(accountType));
      cb.setOnCheckedChangeListener(this);
      mAccountTypesGrid.addView(cb);
    }
    linkInputsWithLabels();
  }

  private @IdRes int getCheckBoxId(AccountType accountType) {
    switch (accountType) {
      case CASH:
        return R.id.AccountTypeCheckboxCash;
      case BANK:
        return R.id.AccountTypeCheckboxBank;
      case CCARD:
        return R.id.AccountTypeCheckboxCcard;
      case ASSET:
        return R.id.AccountTypeCheckboxAsset;
      case LIABILITY:
        return R.id.AccountTypeCheckboxLiability;
    }
    throw new NoSuchElementException();
  }

  protected void saveState() {
    String label = mLabelText.getText().toString();
    if (label.equals("")) {
      mLabelText.setError(getString(R.string.no_title_given));
      return;
    }

    mMethod.setLabel(label);

    mMethod.setPaymentType(mPaymentTypeSpinner.getSelectedItemPosition()-1);
    for (AccountType accountType : AccountType.values()) {
      CheckBox cb = mAccountTypesGrid.findViewWithTag(accountType);
      if (cb.isChecked()) {
        mMethod.addAccountType(accountType);
      } else {
        mMethod.removeAccountType(accountType);
      }
    }
    mMethod.isNumbered = mIsNumberedCheckBox.isChecked();
    //EditActivity.saveState calls DbWriteFragment
    super.saveState();
  }
  @Override
  public Model getObject() {
    return mMethod;
  }
  @Override
  public void onPostExecute(Uri result) {
    setResult(RESULT_OK);
    finish();
    //no need to call super after finish
  }
  protected void setupListeners() {
    mLabelText.addTextChangedListener(this);
    mPaymentTypeSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
      @Override
      public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        setDirty();
      }

      @Override
      public void onNothingSelected(AdapterView<?> parent) {

      }
    });
    mIsNumberedCheckBox.setOnCheckedChangeListener(this);
  }

  @Override
  public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
    setDirty();
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/activity/MethodEdit.kt
/*   This file is part of My Expenses.
 *   My Expenses is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   My Expenses is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with My Expenses.  If not, see <http://www.gnu.org/licenses/>.
*/
package org.totschnig.myexpenses.activity

import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.CompoundButton
import androidx.activity.viewModels
import androidx.appcompat.widget.AppCompatCheckBox
import androidx.core.view.children
import androidx.core.view.isVisible
import androidx.lifecycle.lifecycleScope
import com.evernote.android.state.State
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.databinding.OneMethodBinding
import org.totschnig.myexpenses.dialog.IconSelectorDialogFragment
import org.totschnig.myexpenses.dialog.OnIconSelectedListener
import org.totschnig.myexpenses.injector
import org.totschnig.myexpenses.model.AccountType
import org.totschnig.myexpenses.model.PreDefinedPaymentMethod
import org.totschnig.myexpenses.model2.PaymentMethod
import org.totschnig.myexpenses.provider.DatabaseConstants
import org.totschnig.myexpenses.ui.SpinnerHelper
import org.totschnig.myexpenses.viewmodel.MethodViewModel
import org.totschnig.myexpenses.viewmodel.data.IIconInfo

private const val KEY_SELECTED_TYPES = "selectedTypes"

class MethodEdit : EditActivity(), CompoundButton.OnCheckedChangeListener, OnIconSelectedListener {
    lateinit var binding: OneMethodBinding
    private lateinit var typeSpinner: SpinnerHelper

    private val viewModel by viewModels<MethodViewModel>()

    private val rowId: Long
        get() = intent.extras?.getLong(DatabaseConstants.KEY_ROWID) ?: 0

    @State
    var preDefined: PreDefinedPaymentMethod? = null

    @State
    var icon: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        injector.inject(viewModel)
        binding = OneMethodBinding.inflate(layoutInflater)
        typeSpinner = SpinnerHelper(binding.Type)
        setContentView(binding.root)
        setupToolbarWithClose()
        if (savedInstanceState == null) {
            populateFields()
        } else {
            lifecycleScope.launch {
                setUpAccountTypeGrid(
                    viewModel.accountTypesRaw.first(),
                    savedInstanceState.getLongArray(KEY_SELECTED_TYPES)?.toList()
                )
            }
            linkInputsWithLabels()
        }
        setTitle(if (rowId == 0L) R.string.menu_create_method else R.string.menu_edit_method)
        binding.Icon.setOnClickListener {
            IconSelectorDialogFragment()
                .show(supportFragmentManager, "ICON_SELECTION")
        }
        binding.ClearIcon.setOnClickListener {
            icon = null
            setDirty()
            configureIcon()
        }
        floatingActionButton = binding.fab.CREATECOMMAND
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putLongArray(KEY_SELECTED_TYPES, selectedTypes.toLongArray())
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        setupListeners()
        configureIcon()
    }

    override val fabActionName = "SAVE_METHOD"

    private fun populateFields() {
        if (rowId != 0L) {
            newInstance = false
            lifecycleScope.launch {
                with(viewModel.loadPaymentMethod(rowId)) {
                    binding.Label.setText(label)
                    binding.IsNumbered.isChecked = isNumbered
                    typeSpinner.setSelection(type + 1)
                    preDefined = preDefinedPaymentMethod
                    this@MethodEdit.icon = icon
                    setUpAccountTypeGrid(viewModel.accountTypesRaw.first(), this.accountTypes)
                }
                setupListeners()
                configureIcon()
            }
        } else {
            lifecycleScope.launch {
                setUpAccountTypeGrid(viewModel.accountTypesRaw.first(), emptyList())
            }
            setupListeners()
        }

        linkInputsWithLabels()
    }

    private fun setUpAccountTypeGrid(allTypes: List<AccountType>, checked: List<Long>?) {
        allTypes.forEach { accountType ->
            binding.AccountTypeGrid.addView(AppCompatCheckBox(this).also {
                it.text = accountType.localizedName(this)
                it.isChecked = checked?.contains(accountType.id) == true
                it.tag = accountType.id
                it.setOnCheckedChangeListener(this)
            })
        }
    }

    override fun saveState() {
        val label = binding.Label.text.toString()
        if (label.isEmpty()) {
            binding.Label.error = getString(R.string.required)
        } else {
            isSaving = true
            lifecycleScope.launch {
                viewModel.saveMethod(
                    PaymentMethod(
                        id = rowId,
                        label = binding.Label.text.toString(),
                        icon = icon,
                        type = typeSpinner.selectedItemPosition - 1,
                        isNumbered = binding.IsNumbered.isChecked,
                        accountTypes = selectedTypes,
                        preDefinedPaymentMethod = preDefined
                    )
                )
                finish()
            }
        }
    }

    private val selectedTypes: List<Long>
        get() = binding.AccountTypeGrid.children.mapNotNull {
            (it as? AppCompatCheckBox)?.let { checkBox ->
                if (checkBox.isChecked) checkBox.tag as Long else null
            }
        }.toList()

    private fun setupListeners() {
        binding.Label.addTextChangedListener(this)
        typeSpinner.setOnItemSelectedListener(object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View?,
                position: Int,
                id: Long
            ) {
                setDirty()
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        })
        binding.IsNumbered.setOnCheckedChangeListener(this)
    }

    override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
        setDirty()
    }

    private fun configureIcon() {
        binding.Icon.setCompoundDrawablesRelativeWithIntrinsicBounds(
            icon?.let {
                IIconInfo.resolveIcon(it)?.asDrawable(
                    this, androidx.appcompat.R.attr.colorPrimary
                )
            },
            null, null, null
        )
        binding.ClearIcon.isVisible = icon != null
    }

    override fun onIconSelected(icon: String) {
        setDirty()
        this.icon = icon
        configureIcon()
    }
}